utils/retrieve-state/src/lib.rs

```diff
@@ -289,50 +337,369 @@ pub fn put_block_with_deploys(
+    peers: &[SocketAddr],
+    engine_state: Arc<EngineState<LmdbGlobalState>>,
     state_root_hash: Digest,
-    progress_fn: impl Fn() + Send + Sync,
+    concurrent_requests_per_peer: usize,
 ) -> Result<usize, anyhow::Error> {
-    let mut outstanding_tries = vec![state_root_hash];
+    let (base_send, mut base_recv) = futures_channel::mpsc::unbounded();
+    let mut peer_futures = Vec::new();
+    for peer in peers {
+        let base_send = base_send.clone();
+        let peer_future = async move {
+            let maybe_peer =
+                match maybe_construct_peer(base_send.clone(), *peer, client, state_root_hash).await
+                {
+                    Ok(peer) => Some(peer),
+                    Err(_err) => {
+                        // just skip those that can't be communicated with
+                        // println!("error constructing peer with address {:?} {:?}", peer,
+                        // err);
+                        None
+                    }
+                };
+            maybe_peer
+        };
+        peer_futures.push(peer_future);
+    }
+    let mut peer_map = futures::future::join_all(peer_futures)
+        .await
+        .into_iter()
+        .flatten()
+        .map(|peer: Peer| (peer.address, peer))
+        .collect::<HashMap<SocketAddr, Peer>>();
 
-    let mut start = Instant::now();
-    let mut tries_downloaded = 0;
-    while let Some(next_trie_key) = outstanding_tries.pop() {
-        let read_result = get_trie(
-            client,
-            url,
-            GetTrieParams {
-                trie_key: next_trie_key,
-            },
-        )
-        .await?;
-        if let Some(blob) = read_result.maybe_trie_bytes {
-            let bytes: Vec<u8> = blob.into();
-            let (trie, _): (Trie<Key, StoredValue>, _) = FromBytes::from_bytes(&bytes)?;
-            let mut missing_descendants = engine_state
-                .put_trie_and_find_missing_descendant_trie_keys(CorrelationId::new(), &trie)?;
-            outstanding_tries.append(&mut missing_descendants);
-            tries_downloaded += 1;
-            progress_fn()
-        } else {
-            return Err(anyhow::anyhow!(
-                "unable to download trie at {:?}",
-                next_trie_key
-            ));
+    println!("got {}/{} valid peers", peer_map.len(), peers.len());
+
+    let mut total_tries_count = 0;
+
+    let mut outstanding_trie_keys = vec![state_root_hash];
+    let mut tries_downloaded = 0usize;
+    let mut in_flight_counters = peers
+        .iter()
+        .map(|addr| (addr, 0usize))
+        .collect::<HashMap<_, _>>();
+
+    loop {
+        // Top level task keeps track of missing descendants and work distribution.
+        for (address, peer) in peer_map.iter() {
+            let counter = in_flight_counters.get_mut(&address).unwrap();
+            // TODO: understand what's happening when requests get backed up, causing starvation of
+            // base_recv
```

Comment:
Bit of a show-stopper... I get some serious hiccups because of this

Reply:
Stale comment, not related to the pauses we see where the trie is backed up on writes.